src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1994,6 +2024,129 @@ IntersectPlacementList(List *lhsPlacementList, List *rhsPlacementList)
 
 
 /*
+ * GroupInsertValuesByShardId takes care of grouping the rows from a multi-row
+ * INSERT by target shard. At this point, all pruning has taken place and we
+ * need only to build sets of rows for each destination. This is done by a
+ * simple sort (by shard identifier) and gather step. The sort has the side-
+ * effect of getting things in ascending order to avoid unnecessary deadlocks
+ * during Task execution.
+ */
+static List *
+GroupInsertValuesByShardId(List *insertValuesList)
+{
+	ModifyRoute *route = NULL;
+	ListCell *insertValuesCell = NULL;
+	List *modifyRouteList = NIL;
+
+	insertValuesList = SortList(insertValuesList, CompareInsertValuesByShardId);
```

Comment:
I still think it's worthwhile that if two rows are sorted in a particular order in the initial statement that they maintain that order if sent to the same shard, so fixed it.